DAQ/src/CaloHitsFromFragments_module.cc

```diff
@@ -0,0 +1,327 @@
+    pulseMap_[crystalID].push_back(CrystalInfo(crystalID, 1, time, eDep));
+  }
+}
+
+
+art::CaloHitsFromFragments::CaloHitsFromFragments(const art::EDProducer::Table<Config>& config) :
+  art::EDProducer{config},
+  diagLevel_       (config().diagLevel()), 
+  caloFragmentsTag_(config().caloTag()),
+  digiSampling_    (config().digiSampling()),
+  deltaTPulses_    (config().deltaTPulses()), 
+  pulseRatioMax_   (config().pulseRatioMax()), 
+  pulseRatioMin_   (config().pulseRatioMin()),
+  caloDAQUtil_("CaloHitsFromFragments") {
+    produces<mu2e::CaloHitCollection>();
+    produces<mu2e::CaloHitCollection>("caphri");
+  }
+
+// ----------------------------------------------------------------------
+
+void art::CaloHitsFromFragments::produce(Event& event) {
+  for(size_t i=0; i<pulseMap_.size(); ++i){
+    pulseMap_[i].clear();
+  }
+
+  art::EventNumber_t eventNumber = event.event();
+
+  // Collection of CaloHits for the event
+  std::unique_ptr<mu2e::CaloHitCollection> calo_hits(new mu2e::CaloHitCollection);
+  std::unique_ptr<mu2e::CaloHitCollection> caphri_hits(new mu2e::CaloHitCollection);
+
+  art::Handle<artdaq::Fragments> calFragments;
+  size_t numCalFrags(0);
+  size_t totalSize = 0;
+  event.getByLabel(caloFragmentsTag_, calFragments);
+  if (!calFragments.isValid()) {
+    std::cout << "[CaloHitsFromFragments::produce] found no Calorimeter fragments!"
+	      << std::endl;
+    event.put(std::move(calo_hits));
+    event.put(std::move(caphri_hits));
+    return;
+  }
+  numCalFrags = calFragments->size();
+  for (size_t idx = 0; idx < numCalFrags; ++idx) {
+    auto size = ((*calFragments)[idx]).sizeBytes(); // * sizeof(artdaq::RawDataType);
+    totalSize += size;
+    analyze_calorimeter_((*calFragments)[idx], calo_hits, caphri_hits);
+    //      std::cout << "\tCAL Fragment " << idx << " has size " << size << std::endl;
+  }
+  
+
+  if (diagLevel_ > 1) {
+    std::cout << std::dec << "Producer: Run " << event.run() << ", subrun " << event.subRun()
+              << ", event " << eventNumber << " has " << std::endl;
+    std::cout << numCalFrags << " CAL fragments." << std::endl;
+
+    std::cout << "Total Size: " << (int)totalSize << " bytes." << std::endl;
+  }
+
+  if (diagLevel_ > 0) {
+    std::cout << "mu2e::CaloHitsFromFragments::produce exiting eventNumber="
+              << (int)(event.event()) << " / timestamp=" << (int)eventNumber << std::endl;
+  }
+
+  // Store the calo hits in the event
+  event.put(std::move(calo_hits));
+  event.put(std::move(caphri_hits),"caphri");
+
+
+} // produce()
+
+void art::CaloHitsFromFragments::analyze_calorimeter_(const artdaq::Fragment& f,
+						      std::unique_ptr<mu2e::CaloHitCollection> const& calo_hits,
+						      std::unique_ptr<mu2e::CaloHitCollection> const& caphri_hits) {
+  mu2e::CalorimeterFragment cc(f);
+
+  if (diagLevel_ > 1) {
+    caloDAQUtil_.printCaloFragmentInfo(f, cc);
+  }
+
+  for (size_t curBlockIdx = 0; curBlockIdx < cc.block_count(); curBlockIdx++) {
```

Comment:
Can this part not go in the utility so its not duplicated from StrawAndCaloDigisFromFragments?

Reply:
The operations inside the loop are different. Both modules are looping over the same object, but they produce different outputs that need different processing.

Reply:
I see that I jumped the gun on the merge.  Sorry about that.  If another small PR is warranted I will watch for it.

I am not sure that I understand.  I think that Richie is talking about the call to the print utility at line 206 but Giani is talking about code in the loop that starts at 209?  Can you straighten me out?

Reply:
We're talking about the loop - we'll just have to be careful to keep these both in sync